lib/puppet/functions/foreman/foreman.rb

```diff
@@ -0,0 +1,141 @@
+# Query Foreman
+#
+# The foreman() parser takes various parameters to execute a query against the foreman-api.
+#
+# This sample contains parameters to let us get a list of 'hosts' that match our search parameters.
+#
+#        item         => 'hosts'
+#        search       => 'hostgroup=Grid'
+#        per_page     => '20'
+#        foreman_url  => 'https://foreman'
+#        foreman_user => 'my_api_foreman_user'
+#        foreman_pass => 'my_api_foreman_pass'
+#
+# 'item' may be: environments, fact_values, hosts, hostgroups, puppetclasses, smart_proxies, subnets
+# 'search' is your actual search query.
+# 'per_page' specifies the maximum number of results you'd like to receive.
+#            This defaults to '20' which is consistent with what you'd get from
+#            Foreman if you didn't specify anything.
+# 'foreman_url' is your actual foreman server address
+# 'foreman_user' is the username of an account with API access
+# 'foreman_pass' is the password of an account with API access
+# 'filter_result' string or array with attribites to return
+#                 if a string is given foreman() returns an array only with given attributes
+#                 in case of an array is given foreman() returns an array of hashes selecting only
+#                 attributes given in array
+#                 in case of an given hash foreman() returns an array of hashes selecting only
+#                 attribute keys given in hash renamed to values of given keys. This can be used
+#                 to rename keys in result
+# 'timeout' is the Foreman request timeout in seconds as an integer.
+#           This defaults to five seconds.
+#
+# Then, use a variable to capture its output:
+# $f = ['hosts', 'hostgroup=Grid', '20', 'https://foreman', 'my_api_foreman_user', 'my_api_foreman_pass']
+# $hosts = foreman(*$f)
+#
+# Note: If you're using this in a template, you may be receiving an array of
+# hashes. So you might need to use two loops to get the values you need.
+#
+# Happy Foreman API-ing!
+
+require "yaml"
+require "net/http"
+require "net/https"
+require "uri"
+require "timeout"
+
+Puppet::Functions.create_function(:'foreman::foreman') do
+  dispatch :foreman do
+    required_param 'String', :item
+    required_param 'String', :search
+    optional_param 'Variant[Integer, String]', :per_page
+    optional_param 'String', :foreman_url
```

Comment:
There is a [Stdlib::Httpurl](https://github.com/puppetlabs/puppetlabs-stdlib#stdlibhttpurl)